XDecorations/resources/man/xdecorations.1
.TH "XDECORATIONS" "1" "0.1.10" "K.D.Hedger" "User Commands"
.SH "NAME"
xdecorations - manual page for xdecorations 0.1.9
.br
.SH "DESCRIPTION"
Some eye candy for the holidays, including lights, trees, snowmen, pumpkins, flying sleighs and witches, snow and autumn leaves.
.br

This is along the lines of xsnow/xpenguins, and allows you to set themeable decorations on the desktop, you can choose ( at the moment ) Halloween, Xmas, Autumn or Weather themes or you can add your own.
.br
You can set the speed of the various parts of the theme and their positions, either from the command line or from a config file, ( there is a documented default config file included ).
.br

This works with a number of different WM's, but please see the INSTALL file for details on building/running and some work arounds.
.br

Please see the INSTALL file for basic usage and the documented .rc config file.
.br

Got to my webpage to see some short movies of XDecorations in action with config files.
.br
.SH "MAIN OPTIONS"
These are the application options.
.br

\fB-theme \fRSTRING
.br
	Set theme name ( default 'Xmas' )
.br

\fB-delay \fRINTEGER
.br
	Set main delay
.br

\fB-configfile \fRFILEPATH
.br
	Set new config file ( using only one instance of this makes any sense )
.br

\fB-writeconfig \fRFILEPATH
.br
	Write out a new config file including currently loaded config file,defaults and command line options then exit Options after this are ignored
.br

\fB-noconfig\fR
.br
	Dont't load the default config file ( ~/.config/xdecorations.rc )
.br

\fB-usewindow\fR/-no-usewindow
.br
	Use a transparent window instead of the root window
.br

\fB-usebuffer\fR/-no-usebuffer
.br
	Use double buffering Using double buffering gives best graphical results but is slower The default is to use a transparent window with double buffering
.br

\fB-watchconfig\fR/-no-watchconfig
.br
	Monitor changes to last loaded config file and apply any changes in real time ( experimental be warned! )
.br

\fB-data \fRFOLDERPATH
.br
	Set path to folder containing pixmap data, default is PREFIX/share/XDecorations/pixmaps.
.br

.SH "LAMPS OPTIONS"
These are the options for controlling the lamps.
.br

\fB-lampset\fR INTEGER
.br
	The number of the lamp set to use ( 0=no lamps )
.br

\fB-lampy\fR INTEGER
.br
	Lamp Y position
.br

\fB-lampdelay\fR INTEGER
.br
	Delay for flashing lamps
.br

\fB-lampflash\fR INTEGER
.br
	Lamp flash type 0=flash, 1=inverted chase, 2=chase ,3=random, 4=cycle
.br

\fB-lampcycledelay\fR INTEGER
.br
	Delay before cycling to the next flash type
.br
.SH "FLYERS OPTIONS"
These are the options for controlling the flyers.
.br

\fB-flyer\fR INTEGER
.br
	Flying object set ( 0=disable flyers )
.br

\fB-maxflyers\fR INTEGER
.br
	Number of flying objects to use <100 If the number of flyers are > 0 then the image is chosen randomly from the available flyers If the number of flyers are < 0 then tthe image is chosen sequentially
.br

\fB-flyermaxy\fR INTEGER
.br
	Lowest point on screen for flying objects
.br

\fB-spread\fR INTEGER
.br
	Random delay for flying objects
.br

\fB-flydelay\fR INTEGER
.br
	Flying objects delay
.br

\fB-flystep\fR INTEGER
.br
	Amount to move flying objects
.br

\fB-flyanimdelay\fR INTEGER
.br
	Delay for animation of flying objects
.br
.SH "TREES OPTIONS"
These are the options for controlling the trees.
.br

\fB-tree\fR INTEGER
.br
	The number of the tree to use ( 0=no tree ) If no trees are shown star/treelights/tinsel are also disabled
.br

\fB-treex\fR INTEGER
.br
	Absolute X position of tree
.br

\fB-treey\fR INTEGER
.br
	Absolute Y position of tree
.br

\fB-showstar\fR/\fB-no-showstar\fR
.br
	Show star
.br

\fB-stardelay\fR INTEGER
.br
	Delay for star
.br

\fB-showtinsel\fR/\fB-no-showtinsel\fR
.br
	Show tinsel
.br

\fB-treelampset\fR INTEGER
.br
	Lampset to use on tree
.br

\fB-treelampdelay\fR INTEGER
.br
	Tree lamps delay
.br

You can also use the terms left/centre/right on the command line with \fB-treex\fR to position the tree.
.br
You can also use the terms top/centre/bottom ont the commandline with \fB-treey\fR to position the tree.
.br
To use top, left etc in the config file substitute the following values, LEFT=-1000,CENTRE=-2000,RIGHT=-3000,TOP=-4000,BOTTOM=-5000
.br
.SH "FIGURE OPTIONS"
These are the options for controlling the figure.
.br

\fB-figure\fR INTEGER
.br
	The number of figure to use ( 0=no figure )
.br

\fB-figurex\fR INTEGER
.br
	Absolute X position of figure
.br

\fB-figurey\fR INTEGER
.br
	Absolute Y position of figure
.br

\fB-figuredelay\fR INTEGER
.br
	Delay for figure
.br

You can also use the terms left/centre/right on the command line with \fB-figurex\fR to position the figure.
.br
You can also use the terms top/centre/bottom ont the commandline with \fB-figurey\fR to position the figure.
.br
To use top, left etc in the config file substitute the following values, LEFT=-1000,CENTRE=-2000,RIGHT=-3000,TOP=-4000,BOTTOM=-5000
.br
.SH "FALLING OPTIONS"
These are the options for controlling the falling objects.
.br

\fB-falling\fR INTEGER
.br
	The set number of the falling objects to use ( 0=no falling objects )
.br

\fB-falldelay\fR INTEGER
.br
	Delay for falling objects
.br

\fB-maxfalling\fR INTEGER
.br
	Maximum number of falling objects INTEGER<5000
.br

\fB-fallingspread\fR INTEGER
.br
	Random deleay between new falling objects appearing
.br

\fB-fallingspeed\fR INTEGER
.br
	Distance falling objects move in Y direction in one go
.br

\fB-minfallspeed\fR INTEGER
.br
	Min distance falling objects move in Y direction in one go ( MUST be less than fallingspeed )
.br

\fB-drift\fR INTEGER
.br
	Falling objects max X step
.br

\fB-fallinanimdelay\fR INTEGER
.br
	Delay for animation of falling objects
.br
.SH "WIND OPTIONS"
These are the options for controlling the wind.
.br

\fB-wind\fR INTEGER
.br
	Wind speed/direction ( +INTEGER wind>>> -INTEGER wind<<< 0 no wind )
.br

\fB-usegusts\fR/\fB-no-usegusts\fR
.br
	Use/don't use gusts of wind
.br

\fB-gustlen\fR INTEGER
.br
	Duration of gusts of wind
.br

\fB-gustdelay\fR INTEGER
.br
	Random delay between gusts of wind
.br

\fB-gustspeed\fR INTEGER
.br
	Speed of gusts of wind
.br

\fB-bsettleheight\fR INTEGER
.br
	Height of settled snow etc on bottom of desktop, 0 = Don't let snow settle
.br

\fB-wsettleheight\fR INTEGER
.br
	Height of settled snow etc on windows, 0 = Don't let snow settle
.br

\fB-settlerate\fR INTEGER
.br
	Rate at which settling occurs - lower the number slower but better Settled snow etc is swept clean at the end of a gust You should set drift > 1 for a more relalistic settle after being swept clean
.br

\fB-clearonmaxheight\fR/\fB-no-clearonmaxheight\fR
.br
	Clear the snow etc when it reaches the maximum height set by bsettleheight/wsettleheight
.br

As some window managers include the decorations in the reported window size and some don't you may need to add an offset to where the snow etc starts to settle, KDE will need this xfce4wm may or may not need this.  The bottom offset is to raise the start of settled snow etc from the bottom of the screen if you have a bottom panel.
.br

\fB-woffsety\fR INTEGER
.br
	Offset the settled layer on windows by INTEGER
.br

\fB-boffsety\fR INTEGER
.br
	Offset the settled layer froom the bottom by INTEGER
.br
.SH "AUTHORS"
K.D.Hedger.
.br
keithdhedger@gmail.com
.br
Bugs etc to the above email address.
.br
.SH "SEE ALSO"
Commented example configfile installed in $(pkgdatadir)/docs ( usually /usr/share/XDecorations/docs ).
.br


ChangeLog
0.1.10
Added falling pumpkins to Halloween theme.
Added gpl txt to $(pkgdatadir)/docs.
Added example config file to $(pkgdatadir)/docs.
Fixed some info errors in manpage.
Updated manpage.

0.1.9
Added manpage.
Added path to pixmaps folder to config and options.
moved to autotools project.
Re-arranged layers so that flyers trees and figures don't appear in front of settled snow.

0.1.8
Lamps now in front of falling.
Fixed lampflash number not being updated when using watchconfig.
Code clean.
Added strip command to Makefile.
Added CFLAGS/PREFIX to Makefile, commandline can override.
Snow on windows now in front of figures and flyers.

0.1.7
Installed INSTALL file to /usr/share/doc/XDecorations/.
This is the last release that will write to the root window as most distros seem to block this one way or another.
Settling snow etc does NOT work on ubuntu unity but is fine for xfce,kde,mint gnome,lxde.
Added y offsets for snow etc settling on windows.
Added y offsets for snow etc settling on bottom.
Added packaging script thanks to Mario for that.
Added different max settling height to windows.
Changed some option names.
Added falling settling on windows.
Split source code into seperate files.
Added falling settling at bottom of screen.

0.1.7
Added 'retro' xsnow snow, for us old timer's.
Installed example config to /usr/share/doc/XDecorations/xdecorations.rc.
Added doc folder to /usr/share/doc.
Falling snow etc now settles at bottom of window. 

0.1.6
Added new snowmen.
Added 2 different size santas.
Added new Xmas tree lights.
Added 2 new Xmas trees.

0.1.5
Can now change figure/tree/lamps variables from watched configfile.
Can now change falling/flying variables from watched configfile.
falling objects  speed etc re randomized after reaching bottom.
Started to add watch config.
Can now set flyer image chosen from random loaded flyers ( default ) or sequentially ( use negative number ).
Can now set the number of flying objects.
Added clouds to Autumn.
Can now write config file to stdout ( use - as  the argument to writeconfig ).
Added 'Weather' theme.
Added -minfallspeed to options.
Added rain to Autumn ( You can tell I live in England! ).
Falling animations now go either backward or fowards randomly.
Gusts now affect flyers.
Added two extra lamp sets to xmas theme.
Added different flash types to lamps.
Added animations to flying objects.
Randomized anim start position of falling objects.

0.1.4
Split Themes into seperate folders.
Added animations for falling objects.
added -noconfig to options to not load the default config file ~/.config/xdecorations.rc
Changed -holiday to -theme.
Added gusting wind.
Added double buffering.
Added theme "Autumn"
Added falling objects.

0.1.3
Default is now to use a transparent window instead of writing to the root window.
Fixed tinsel bug.
Added -no-usewindow option to overide config file/default setting.
Added true transparency, needs to be run in window with composite manager.
Changed image format from XPM to PNG.
Added -usewindow option, to use big transparent window instead of writing to root window.
Added imlib2 dependency.

0.1.2
Fixed not reloading config file properly before using -writeconfig.
Now no limits on number of lamps.
Now no limits on number of figures.
Now no limits on number of tree lamps.
Now no limits on number of trees.
Updated -help.
Added -writeconfig to command line to write out a new config file including currently loaded config file,defaults and commandline options.
Removed unneeded sanity check.
Now only loads the needed xpm files.

0.1.1
Added INSTALL file.
Added out of bounds check for figure number/tree number/lamp number.
Added -configfile option to command line to use a different config file.
Added -figurenumber to command line and config file.
Added multiple figures.
Added -lampset option to command line and config file.
Added multiple lamps.

0.1.0
First release.

xdecorations.c
#if 0
©Keith Hedger Mon 20 Oct 14:21:11 BST 2014 kdhedger68713@gmail.com
*   This program is free software;
you can redistribute it and/or modify
*   it under the terms of the GNU General Public License as published by
*   the Free Software Foundation;
either version 2 of the License,or
*   (at your option) any later version.
*
*   This program is distributed in the hope that it will be useful,
*   but WITHOUT ANY WARRANTY;
without even the implied warranty of
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*   GNU General Public License for more details.
*
*   You should hargve received a copy of the GNU General Public License
*   along with this program;
if not,write to the Free Software
*   Foundation,Inc.,59 Temple Place,Suite 330,Boston,MA  02111-1307  USA

*   toon_root.c Copyright (C) 1999-2001  Robin Hogan
#endif

#include <X11/Intrinsic.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xos.h>
#include <X11/xpm.h>

#include <stdio.h>
#include <signal.h>
#include <stdlib.h>

#define MAXPATHNAMELEN 2048
#define NUMBEROFFLYERS 2

char		pathname[MAXPATHNAMELEN];

Display*	display;
Window		rootWin;
int			displayWidth;
int			displayHeight;
int			centerX;
int			centerY;
GC			gc;

int			done=0;
long		mainDelay=50000;

Pixmap		figurePixmap[2];
Pixmap		figureMaskPixmap[2];

Pixmap		lampsPixmap[2];
Pixmap		lampsMaskPixmap[2];

Pixmap		treePixmap[2];
Pixmap		treeMaskPixmap[2];

Pixmap		starPixmap[2];
Pixmap		starMaskPixmap[2];

Pixmap		treeLampsPixmap[8];
Pixmap		treeLampsMaskPixmap[8];

Pixmap		tinselPixmap;
Pixmap		tinselMaskPixmap;

Pixmap		flyersPixmap[NUMBEROFFLYERS];
Pixmap		flyersMaskPixmap[NUMBEROFFLYERS];

int			onOff=0;

uint		runCounter=0;

int			flyersSpeed=1;
int			flyersStep=10;
int			flyersWidth[NUMBEROFFLYERS];
int			flyersHeight[NUMBEROFFLYERS];
int			flyersX[NUMBEROFFLYERS];
int			flyersY[NUMBEROFFLYERS];
int			showFlyers=1;
int			flyersMaxY=400;
int			flyersActive[NUMBEROFFLYERS];
int			flyerSpread=400;

int			figureSpeed=10;
int			figureX=100;
int			figureY=100;
int			figureW;
int			figureH;
int			figureCount=0;
int			figure=0;
int			figureOnOff=0;

int			lampSpeed=10;
int			lampX=0;
int			lampY=26;
int			lampWidth;
int			lampHeight;
int			lampCount=0;
int			showLamps=1;
int			lampOffset=0;

int			treeWidth;
int			treeHeight;
int			treeNumber=1;
int			treelampSpeed=8;
int			starSpeed=6;
int			treeX=100;
int			treeY=100;
int			treeLampSet=1;
int			showTinsel=1;
int			showStar=1;
int			showTree=1;
int			showTreeLamps=1;

int			treeOnOff=0;
int			starOnOff=0;

const char*	prefix="Xmas";

//struct args
//{
//	const char*	name;
//	int			type;
//	void*		data;
//};
//
//enum {TYPEINT=1,TYPESTRING,TYPEBOOL};
//
//int indent=0;
//
//args			kkedit_rc[]=
//{
//	//bools
//	{"indentcode",TYPEBOOL,&indent},
//	{NULL,0,NULL}
//};

//void loadVarsFromFile(char* filepath,args* dataptr)
//{
//	FILE*	fd=NULL;
//	char	buffer[2048];
//	int		cnt;
//	char*	argname=NULL;
//	char*	strarg=NULL;
//
//	fd=fopen(filepath,"r");
//	if(fd!=NULL)
//		{
//			while(feof(fd)==0)
//				{
//					buffer[0]=0;
//					fgets(buffer,2048,fd);
//					sscanf(buffer,"%as %as",&argname,&strarg);
//					cnt=0;
//					while(dataptr[cnt].name!=NULL)
//						{
//							if((strarg!=NULL) && (argname!=NULL) && (strcmp(argname,dataptr[cnt].name)==0))
//								{
//									switch(dataptr[cnt].type)
//										{
//											case TYPEINT:
//												*(int*)dataptr[cnt].data=atoi(strarg);
//												break;
//											case TYPESTRING:
//												if(*(char**)(dataptr[cnt].data)!=NULL)
//													free(*(char**)(dataptr[cnt].data));
//												sscanf(buffer,"%*s %a[^n]s",(char**)dataptr[cnt].data);
//												break;
//											case TYPEBOOL:
//												*(bool*)dataptr[cnt].data=(bool)atoi(strarg);
//												break;
//										}
//								}
//							cnt++;
//						}
//					if(argname!=NULL)
//						free(argname);
//					if(strarg!=NULL)
//						free(strarg);
//					argname=NULL;
//					strarg=NULL;
//				}
//			fclose(fd);
//		}
//}

void signalHandler()
{
	done=1;
}

int randInt(int maxVal)
{
	return rand() % maxVal;
}

void uSsleep(unsigned long usec)
{
	struct timeval t;
	t.tv_usec=usec%(unsigned long)1000000;
	t.tv_sec=usec/(unsigned long)1000000;
	select(0,(void* )0,(void* )0,(void* )0,&t);
}

void initFlyers(void)
{
	int				rc=0;
	int				j=0;
	XpmAttributes	attrib;

	attrib.valuemask=0;

	rc=0;
	for(j=0; j<NUMBEROFFLYERS;j++)
		{
			snprintf(pathname,MAXPATHNAMELEN,"%s/%sFly%i.xpm",DATADIR,prefix,j+1);
			rc+=XpmReadFileToPixmap(display,rootWin,pathname,&flyersPixmap[j],&flyersMaskPixmap[j],&attrib);
			flyersWidth[j]=attrib.width;
			flyersHeight[j]=attrib.height;
			flyersY[j]=(rand() % flyersMaxY);
			flyersX[j]=10;
			if((rand() % flyerSpread)==0)
				flyersActive[j]=1;
			else
				flyersActive[j]=0;
				
		}
	if(rc!=0)
		showFlyers=0;
}

void initFigure(void)
{
	int				rc=0;
	XpmAttributes	attrib;
	char*			lampseton;
	char*			lampsetoff;

	attrib.valuemask=0;

	asprintf(&lampseton,"%s/%sFigureOn.xpm",DATADIR,prefix);
	asprintf(&lampsetoff,"%s/%sFigureOff.xpm",DATADIR,prefix);

	rc+=XpmReadFileToPixmap(display,rootWin,lampsetoff,&figurePixmap[0],&figureMaskPixmap[0],&attrib);
	rc+=XpmReadFileToPixmap(display,rootWin,lampseton,&figurePixmap[1],&figureMaskPixmap[1],&attrib);
	if(rc!=0)
		figure=0;
	figureW=attrib.width;
	figureH=attrib.height;

	free(lampseton);
	free(lampsetoff);
}

void initLamps(void)
{
	int				rc=0;
	XpmAttributes	attrib;
	char*			lampseton;
	char*			lampsetoff;

	attrib.valuemask=0;

	asprintf(&lampseton,"%s/%sLampsOn.xpm",DATADIR,prefix);
	asprintf(&lampsetoff,"%s/%sLampsOff.xpm",DATADIR,prefix);

	rc+=XpmReadFileToPixmap(display,rootWin,lampsetoff,&lampsPixmap[0],&lampsMaskPixmap[0],&attrib);
	rc+=XpmReadFileToPixmap(display,rootWin,lampseton,&lampsPixmap[1],&lampsMaskPixmap[1],&attrib);
	if(rc!=0)
		showLamps=0;
	lampWidth=attrib.width;
	lampHeight=attrib.height;

	lampCount=(displayWidth/lampWidth)+1;
	lampOffset=(displayWidth-(lampWidth*lampCount))/2;

	free(lampseton);
	free(lampsetoff);
}

void initTree(void)
{
	int				rc=0;
	int				j=0;
	XpmAttributes	attrib;

	attrib.valuemask=0;

	snprintf(pathname,MAXPATHNAMELEN,"%s/%sTree1.xpm",DATADIR,prefix);
	rc+=XpmReadFileToPixmap(display,rootWin,pathname,&treePixmap[0],&treeMaskPixmap[0],&attrib);
	snprintf(pathname,MAXPATHNAMELEN,"%s/%sTree2.xpm",DATADIR,prefix);
	rc+=XpmReadFileToPixmap(display,rootWin,pathname,&treePixmap[1],&treeMaskPixmap[1],&attrib);
	if(rc!=0)
		showTree=0;

	treeWidth=attrib.width;
	treeHeight=attrib.height;

	rc=0;
	for(j=0; j<8; j++)
		{
			snprintf(pathname,MAXPATHNAMELEN,"%s/%sTreeLights%i.xpm",DATADIR,prefix,j+1);
			rc+=XpmReadFileToPixmap(display,rootWin,pathname,&treeLampsPixmap[j],&treeLampsMaskPixmap[j],&attrib);

		}
	if(rc!=0)
		showTreeLamps=0;

	rc=0;
	snprintf(pathname,MAXPATHNAMELEN,"%s/%sStar0.xpm",DATADIR,prefix);
	rc+=XpmReadFileToPixmap(display,rootWin,pathname,&starPixmap[0],&starMaskPixmap[0],&attrib);
	snprintf(pathname,MAXPATHNAMELEN,"%s/%sStar1.xpm",DATADIR,prefix);
	rc+=XpmReadFileToPixmap(display,rootWin,pathname,&starPixmap[1],&starMaskPixmap[1],&attrib);
	if(rc!=0)
		showStar=0;

	rc=0;
	rc+=XpmReadFileToPixmap(display,rootWin,DATADIR "/Tinsel.xpm",&tinselPixmap,&tinselMaskPixmap,&attrib);
	if(rc!=0)
		showTinsel=0;
}

void drawFlyers(void)
{
	int rc;
	int	j;

	Region	flyreg=XCreateRegion();
	Region	final;
	XRectangle AddRect;

//	for(j=0;j<NUMBEROFFLYERS;j++)
//		{
//			rc=XClearArea(display,rootWin,flyersX[j],flyersY[j],flyersWidth[j],flyersHeight[j],False);
////			AddRect.x=flyersX[j];
////			AddRect.y=flyersY[j];
////			AddRect.width=flyersWidth[j];
////			AddRect.height=flyersHeight[j];
////			
////			final=XUnionRectWithRegion(AddRect,flyreg);
//		}
//	rc=XClearArea(display,rootWin,flyersX[j],flyersY[j],flyersWidth[j],flyersHeight[j],False);

	if (showFlyers==1)
		{
			for(j=0;j<NUMBEROFFLYERS;j++)
				{
					if(flyersActive[j]==1)
						{
							rc=XSetClipMask(display,gc,flyersMaskPixmap[j]);
							rc=XSetClipOrigin(display,gc,flyersX[j],flyersY[j]);
	//						rc=XClearArea(display,rootWin,flyersX[j],flyersY[j],flyersWidth[j],flyersHeight[j],False);
							rc=XCopyArea(display,flyersPixmap[j],rootWin,gc,0,0,flyersWidth[j],flyersHeight[j],flyersX[j],flyersY[j]);
						}
				}
		}
}

void drawFigure(void)
{
	int rc;
	if (figure==1)
		{
			rc=XSetClipMask(display,gc,figureMaskPixmap[figureOnOff]);
			rc=XSetClipOrigin(display,gc,figureX,figureY);
//			rc=XClearArea(display,rootWin,figureX,figureY,figureW,figureH,False);
			rc=XCopyArea(display,figurePixmap[figureOnOff],rootWin,gc,0,0,figureW,figureH,figureX,figureY);
		}
}

void drawLamps(void)
{
	int rc;
	int loop;
	int	CurrentlampX=lampX;
	rc=XSetClipMask(display,gc,lampsMaskPixmap[onOff]);

	for (loop=0; loop<lampCount; loop++)
		{
			rc=XSetClipOrigin(display,gc,CurrentlampX+lampOffset,lampY);
			rc=XCopyArea(display,lampsPixmap[onOff],rootWin,gc,0,0,lampWidth,lampHeight,CurrentlampX+lampOffset,lampY);
			CurrentlampX+=lampWidth;
		}
}

updateLamps(void)
{
	onOff=(onOff+1) & 1;
}

void drawTreeLamps(void)
{
	int rc;
	int lastlamp;
	int lampset;

	if (treeLampSet==2)
		lampset=4;
	else
		lampset=0;

	rc=XSetClipMask(display,gc,treeMaskPixmap[treeNumber-1]);
	rc=XSetClipOrigin(display,gc,treeX,treeY);
//	rc=XClearArea(display,rootWin,treeX,treeY,treeWidth,treeHeight,False);
	rc=XCopyArea(display,treePixmap[treeNumber-1],rootWin,gc,0,0,treeWidth,treeHeight,treeX,treeY);

	if(showTreeLamps==1)
		{
			rc=XSetClipMask(display,gc,treeLampsMaskPixmap[treeOnOff+lampset]);
			rc=XCopyArea(display,treeLampsPixmap[treeOnOff+lampset],rootWin,gc,0,0,treeWidth,treeHeight,treeX,treeY);
		}

	if (showStar==1)
		{
			rc=XSetClipMask(display,gc,starMaskPixmap[starOnOff]);
			rc=XCopyArea(display,starPixmap[starOnOff],rootWin,gc,0,0,treeWidth,treeHeight,treeX,treeY);
		}

	if (showTinsel==1)
		{
			rc=XSetClipMask(display,gc,tinselMaskPixmap);
			rc=XCopyArea(display,tinselPixmap,rootWin,gc,0,0,treeWidth,treeHeight,treeX,treeY);
		}
}

void updateTreeLamps(void)
{
	treeOnOff++;

	if (treeOnOff==4)
		treeOnOff=0;
}

void updateStar(void)
{
	starOnOff=(starOnOff+1) & 1;
}

void updateFigure(void)
{
	figureOnOff=(figureOnOff+1) & 1;
}

void updateFlyers(void)
{
	int	j=0;

	for(j=0;j<NUMBEROFFLYERS;j++)
		{
			if((flyersActive[j]==0) && ((rand() % flyerSpread)==0))
				flyersActive[j]=1;

			if(flyersActive[j]==1)
				{
					flyersX[j]+=flyersStep;

					if(flyersX[j]>displayWidth)
						{
							flyersActive[j]=0;
							flyersX[j]=0;
							flyersY[j]=(rand() % flyersMaxY);
						}
				}
		}
}

void eraseRects(void)
{
	int	rc;
	int	j;

	if (showTree==1)
		{
			rc=XClearArea(display,rootWin,treeX,treeY,treeWidth,treeHeight,False);
		}

	if (figure==1)
		{
			rc=XClearArea(display,rootWin,figureX,figureY,figureW,figureH,False);
		}

	if (showFlyers==1)
		{
			for(j=0;j<NUMBEROFFLYERS;j++)
				rc=XClearArea(display,rootWin,flyersX[j],flyersY[j],flyersWidth[j],flyersHeight[j],False);
		}

}

int main(int argc,char* argv[])
{
	int argnum;
	char* arg;
	XEvent ev;
	Window root;
	int screen;
	Window		parentWindow;
//command line options.
	for (argnum=1; argnum<argc; argnum++)
		{
			arg=argv[argnum];


			if (strcmp(arg,"-speed")==0)
				mainDelay=atol(argv[++argnum]);
	
			if (strcmp(arg,"-flyermaxy")==0)
				flyersMaxY=atol(argv[++argnum]);

			if (strcmp(arg,"-spread")==0)
				flyerSpread=atol(argv[++argnum]);

			if (strcmp(arg,"-noflyer")==0)
				showFlyers=0;

			if (strcmp(arg,"-lampspeed")==0)
				lampSpeed=atol(argv[++argnum]);

			if (strcmp(arg,"-flyspeed")==0)
				flyersSpeed=atol(argv[++argnum]);

			if (strcmp(arg,"-flystep")==0)
				flyersStep=atol(argv[++argnum]);

			if (strcmp(arg,"-lampy")==0)
				lampY=atol(argv[++argnum]);

			if (strcmp(arg,"-treelampspeed")==0)
				treelampSpeed=atol(argv[++argnum]);

			if (strcmp(arg,"-treelampset")==0)
				treeLampSet=atol(argv[++argnum]);

			if (strcmp(arg,"-treenumber")==0)
				treeNumber=atol(argv[++argnum]);

			if (strcmp(arg,"-treex")==0)
				treeX=atol(argv[++argnum]);

			if (strcmp(arg,"-treey")==0)
				treeY=atol(argv[++argnum]);

			if (strcmp(arg,"-starspeed")==0)
				starSpeed=atol(argv[++argnum]);

			if (strcmp(arg,"-noxmastree")==0)
				showTree=0;

			if (strcmp(arg,"-nolamps")==0)
				showLamps=0;

			if (strcmp(arg,"-notreelamps")==0)
				showTreeLamps=0;

			if (strcmp(arg,"-nostar")==0)
				showStar=0;

			if (strcmp(arg,"-notinsel")==0)
				showTinsel=0;

			if (strcmp(arg,"-holiday")==0)
				prefix=argv[++argnum];

			if (strcmp(arg,"-figure")==0)
				figure=1;

			if (strcmp(arg,"-figurex")==0)
				figureX=atol(argv[++argnum]);

			if (strcmp(arg,"-figurey")==0)
				figureY=atol(argv[++argnum]);

			if (strcmp(arg,"-figurespeed")==0)
				figureSpeed=atol(argv[++argnum]);
		}

	srand((int)time((long* )NULL));

	signal(SIGKILL,signalHandler);
	signal(SIGTERM,signalHandler);
	signal(SIGQUIT,signalHandler);
	signal(SIGINT,signalHandler);
	signal(SIGHUP,signalHandler);

	display=XOpenDisplay(NULL);
	if (display==NULL)
		exit(1);

	screen=DefaultScreen(display);
	rootWin=ToonGetRootWindow(display,screen,&parentWindow);

	displayWidth=DisplayWidth(display,screen);
	displayHeight=DisplayHeight(display,screen);
	centerX=displayWidth / 2;
	centerY=displayHeight / 2;

	initLamps();
	initTree();
	initFigure();
	initFlyers();

	gc=XCreateGC(display,rootWin,0,NULL);
	XSetFillStyle(display,gc,FillStippled);

	XSelectInput(display,rootWin,ExposureMask | SubstructureNotifyMask);

	while (!done)
		{
			while (XPending(display))
				{
					XNextEvent(display,&ev);

				}
			uSsleep(mainDelay);

			runCounter++;

			eraseRects();

			if (showLamps==1)
				{
					if ((runCounter % lampSpeed)==0)
						updateLamps();
					drawLamps();
				}

			if (showTree==1)
				{
					if ((runCounter % treelampSpeed)==0)
						updateTreeLamps();
					drawTreeLamps();

					if ((runCounter % starSpeed)==0 && showStar==1)
						updateStar();
				}

			if (figure==1)
				{
					if ((runCounter % figureSpeed)==0)
						updateFigure();
					drawFigure();
				}

			if (showFlyers==1)
				{
					if ((runCounter % flyersSpeed)==0)
						updateFlyers();
					drawFlyers();
				}

		}
	XClearWindow(display,rootWin);
	XCloseDisplay(display);
	exit(0);
}


